app/jobs/csv_export_job.rb

```diff
@@ -0,0 +1,154 @@
+class CsvExportJob < ActiveJob::Base
+  queue_as :csv_jobs
+ 
+  def perform(csv_export_id)
+    ActiveRecord::Base.connection_pool.with_connection do
+      csv_export = CsvExport.find(csv_export_id)
+      create_export_folder
+      export_task(csv_export)
+      cleanup_downloaded
+    end
+  end
+  
+  private
+  
+  def cleanup_downloaded
+    # clean up downloaded files older than 12 hours or any jobs stuck for over 1 month
+    # TODO delete files with no record associations
+    end_date = Time.now - Rails.application.config.samson.export_job.downloaded_age
+    timeout_date = Time.now - Rails.application.config.samson.export_job.max_age
+    @csv_exports = CsvExport.where("(status = 'downloaded' AND updated_at <= :end_date) OR updated_at <= :timeout_date",
+      {end_date: end_date, timeout_date: timeout_date})
+    if @csv_exports
+      @csv_exports.each do |csv_downloaded|
+        filename = csv_downloaded.full_filename
+        File.delete(filename) unless !File.exist?(filename)
+        csv_downloaded.delete
+        Rails.logger.info("Downloaded file #{csv_downloaded.filename} deleted")
+      end
+    end
+  end
+  
+  def export_task(csv_export)
+    if csv_export.content == "deploys"
+      csv_export.started!
+      deploy_csv_export(csv_export)
+    else
+      csv_export.failed!
+    end
+
+    notify_of_creation(csv_export)
+  rescue
+    #delete_file(csv_export.full_filename)
+    csv_export.failed!
+    Rails.logger.info("Export #{csv_export.filename} failed")
+  end
+  
+  def notify_of_creation(csv_export)
+    if csv_export.status? :started
+      CsvMailer.created_email(csv_export).deliver_now unless (csv_export.email.nil? || csv_export.email.empty?)
+      csv_export.finished!
+      Rails.logger.info("Export #{csv_export.filename} completed")
+    end
+  end
+  
+  def deploy_csv_export(csv_export)
+    filename = csv_export.full_filename
+    filter = deploy_filter(csv_export)
+
+    @deploys = Deploy.joins(:stage, :job).where(filter)
+    summary = [ "-", "Deploys", @deploys.count.to_s ]
+    filters_applied = [ "-", "Filters", filter.to_json ]
+
+    CSV.open(filename, 'w+') do |csv|
+      csv << ["Deploy Number", "Project Name", "Deploy Sumary", "Deploy Commit", "Deploy Status", "Deploy Updated",
+        "Deploy Created", "Deployer Name" "Deployer Email", "Buddy Name", "Buddy Email",
+        "Production Flag", "Bypass Buddy Check" ]
+      Deploy.uncached do
+        @deploys.find_each do |deploy|
+          csv << [deploy.id, deploy.project.name, deploy.summary, deploy.commit, deploy.job.status, deploy.updated_at,
+            deploy.start_time, deploy.job.user.name, deploy.job.user.try(:email), deploy.csv_buddy, deploy.buddy_email,
+            deploy.stage.production, deploy.stage.bypass_buddy_check ]
+        end
+      end
+      csv << summary
+      csv << filters_applied
+    end
+  end
+  
+  def create_export_folder
+    Dir.mkdir("#{Rails.root}/export") unless File.exist?("#{Rails.root}/export")
```

Comment:
now csv_export and the job know about the export folder -> only csv_export should know -> make it a constant or make it `File.dirname(csv_export.full_filename)`